### PR TITLE
[Core] Fix segfault when cancel and generator is used with a high load

### LINKED ISSUE #39703: [Serve][Streaming] Actor crash when requests get canceled.
### What happened + What you expected to happen

doing locust testing, when click ctrl+c, ray actors crash when cancelling requests.

```
ted during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,649 http_proxy 10.0.24.103 a9db2d2b-652f-459c-bde0-6054d10f8737 / default http_proxy.py:1229 - Client for request a9db2d2b-652f-459c-bde0-6054d10f8737 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,650 http_proxy 10.0.24.103 af4d416d-6e1e-4d59-ad91-a827c4f7f7b2 / default http_proxy.py:1229 - Client for request af4d416d-6e1e-4d59-ad91-a827c4f7f7b2 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,650 http_proxy 10.0.24.103 6702b8b0-59e9-4370-832b-4654369c4140 / default http_proxy.py:1229 - Client for request 6702b8b0-59e9-4370-832b-4654369c4140 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,651 http_proxy 10.0.24.103 f5d9f5ef-e988-4de4-a22b-f4c1f67cce45 / default http_proxy.py:1229 - Client for request f5d9f5ef-e988-4de4-a22b-f4c1f67cce45 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,651 http_proxy 10.0.24.103 c3a6392e-52d3-492c-ab40-5d08e07b26f9 / default http_proxy.py:1229 - Client for request c3a6392e-52d3-492c-ab40-5d08e07b26f9 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,652 http_proxy 10.0.24.103 4994b93d-9803-472f-8a7e-450a88ec322a / default http_proxy.py:1229 - Client for request 4994b93d-9803-472f-8a7e-450a88ec322a disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,652 http_proxy 10.0.24.103 d76f0559-e0bc-45b5-8850-820763b8b08d / default http_proxy.py:1229 - Client for request d76f0559-e0bc-45b5-8850-820763b8b08d disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,653 http_proxy 10.0.24.103 9e1ff4bf-bc8c-48b3-ab67-b414c6370473 / default http_proxy.py:1229 - Client for request 9e1ff4bf-bc8c-48b3-ab67-b414c6370473 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,654 http_proxy 10.0.24.103 35507094-6b51-4bd6-9577-135088521f55 / default http_proxy.py:1229 - Client for request 35507094-6b51-4bd6-9577-135088521f55 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,654 http_proxy 10.0.24.103 ad389201-40f6-4449-99c4-425bdc1ad2bc / default http_proxy.py:1229 - Client for request ad389201-40f6-4449-99c4-425bdc1ad2bc disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,655 http_proxy 10.0.24.103 4d29629b-b737-4a78-bf9c-cedce921412f / default http_proxy.py:1229 - Client for request 4d29629b-b737-4a78-bf9c-cedce921412f disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,655 http_proxy 10.0.24.103 23a9e44f-d74b-4e04-bd92-a54983648058 / default http_proxy.py:1229 - Client for request 23a9e44f-d74b-4e04-bd92-a54983648058 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,656 http_proxy 10.0.24.103 16f270b5-d516-4271-9529-53b0e7dbb5f1 / default http_proxy.py:1229 - Client for request 16f270b5-d516-4271-9529-53b0e7dbb5f1 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,656 http_proxy 10.0.24.103 84fd9a92-40e8-4528-b211-007781879e02 / default http_proxy.py:1229 - Client for request 84fd9a92-40e8-4528-b211-007781879e02 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,657 http_proxy 10.0.24.103 3f21c8d8-23e2-40cf-838a-eb358f1495f2 / default http_proxy.py:1229 - Client for request 3f21c8d8-23e2-40cf-838a-eb358f1495f2 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,657 http_proxy 10.0.24.103 083e0f1e-1882-4d8d-8018-a24cc2ccb422 / default http_proxy.py:1229 - Client for request 083e0f1e-1882-4d8d-8018-a24cc2ccb422 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,658 http_proxy 10.0.24.103 9fabd535-293a-4530-bb00-095c2085475a / default http_proxy.py:1229 - Client for request 9fabd535-293a-4530-bb00-095c2085475a disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,658 http_proxy 10.0.24.103 fbdca48c-0e6e-49c2-80d8-a6480046227d / default http_proxy.py:1229 - Client for request fbdca48c-0e6e-49c2-80d8-a6480046227d disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,659 http_proxy 10.0.24.103 523e42bd-7ca9-43db-adbf-f063cf50d031 / default http_proxy.py:1229 - Client for request 523e42bd-7ca9-43db-adbf-f063cf50d031 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,660 http_proxy 10.0.24.103 63390e6e-4163-403d-862c-564bd57140e6 / default http_proxy.py:1229 - Client for request 63390e6e-4163-403d-862c-564bd57140e6 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,660 http_proxy 10.0.24.103 79cdf044-04a6-4475-a154-7804c89178b5 / default http_proxy.py:1229 - Client for request 79cdf044-04a6-4475-a154-7804c89178b5 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,661 http_proxy 10.0.24.103 39b36f41-f647-45ca-be9c-2093eaa447f2 / default http_proxy.py:1229 - Client for request 39b36f41-f647-45ca-be9c-2093eaa447f2 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,662 http_proxy 10.0.24.103 28c60e16-c315-4e29-9f65-8eeeaa188696 / default http_proxy.py:1229 - Client for request 28c60e16-c315-4e29-9f65-8eeeaa188696 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,662 http_proxy 10.0.24.103 21af5ff1-adbc-4070-8c85-f453edab1729 / default http_proxy.py:1229 - Client for request 21af5ff1-adbc-4070-8c85-f453edab1729 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,663 http_proxy 10.0.24.103 0bdb1d43-1f7e-4d59-b43f-6ed889095f2b / default http_proxy.py:1229 - Client for request 0bdb1d43-1f7e-4d59-b43f-6ed889095f2b disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,663 http_proxy 10.0.24.103 addb1b0a-b9d6-4433-8b42-bdb558f3de63 / default http_proxy.py:1229 - Client for request addb1b0a-b9d6-4433-8b42-bdb558f3de63 disconnected during execution, cancelling request.
(HTTPProxyActor pid=77208) INFO 2023-09-15 14:32:32,665 http_proxy 10.0.24.103 a9582505-8ccb-4792-9007-4033a3de6569 / default http_proxy.py:1229 - Client for request a9582505-8ccb-4792-9007-4033a3de6569 disconnected during execution, cancelling request.
(ServeReplica:default:Router pid=77287) INFO 2023-09-15 14:32:32,606 Router default#Router#HzAuSL 8d07a634-d7c5-40f1-aaa3-975e30f39987 / default replica.py:749 - __CALL__ OK 8825.3ms
(ServeReplica:default:Router pid=77287) INFO 2023-09-15 14:32:32,637 Router default#Router#HzAuSL 83a85461-9ed9-4d26-9aa4-7ae8e73f19ae / default replica.py:749 - __CALL__ OK 6649.9ms
(ServeReplica:default:SimpleGenerator pid=77282) *** SIGSEGV received at time=1694813552 on cpu 30 ***
(ServeReplica:default:SimpleGenerator pid=77282) PC: @     0x7f8b4c1528c7  (unknown)  (unknown)
(ServeReplica:default:SimpleGenerator pid=77282)     @     0x7f8b4c41d420  (unknown)  (unknown)
(ServeReplica:default:SimpleGenerator pid=77282)     @                0x1  (unknown)  (unknown)
(ServeReplica:default:SimpleGenerator pid=77282) [2023-09-15 14:32:32,654 E 77282 77316] logging.cc:361: *** SIGSEGV received at time=1694813552 on cpu 30 ***
(ServeReplica:default:SimpleGenerator pid=77282) [2023-09-15 14:32:32,656 E 77282 77316] logging.cc:361: PC: @     0x7f8b4c1528c7  (unknown)  (unknown)
(ServeReplica:default:SimpleGenerator pid=77282) [2023-09-15 14:32:32,656 E 77282 77316] logging.cc:361:     @     0x7f8b4c41d420  (unknown)  (unknown)
(ServeReplica:default:SimpleGenerator pid=77282) [2023-09-15 14:32:32,658 E 77282 77316] logging.cc:361:     @                0x1  (unknown)  (unknown)
(ServeReplica:default:SimpleGenerator pid=77282) Fatal Python error: Segmentation fault
```

**note:**
1. actor is crashed when I use time.sleep() in the `SimpleGenerator` class. If switch to async sleep, the actor is not crashing.

### Versions / Dependencies

master

### Reproduction script

stream.py
```
from fastapi import FastAPI
from ray import serve
from starlette.responses import StreamingResponse
from starlette.requests import Request
import requests
import time
import asyncio
from data2 import AviaryModelResponse
import time


app = FastAPI()


@serve.deployment(max_concurrent_queries=1000)
@serve.ingress(app)
class Router:
    def __init__(self, handle) -> None:
        self._h = handle.options(stream=True)
        self.total_recieved = 0

    @app.get("/")
    def stream_hi(self, request: Request) -> StreamingResponse:
        async def consume_obj_ref_gen():
            obj_ref_gen = await self._h.hi_gen.remote()
            start = time.time()
            num_recieved = 0
            async for chunk in obj_ref_gen:
                chunk = await chunk
                num_recieved +=1 
                yield str(chunk.json())
            delta = time.time() - start
            print(f"**request throughput: {num_recieved / delta}")
        return StreamingResponse(consume_obj_ref_gen(), media_type="text/plain")


@serve.deployment(max_concurrent_queries=1000)
class SimpleGenerator:
    async def hi_gen(self):
        start = time.time()
        for i in range(100):
            #await asyncio.sleep(0.001)
            time.sleep(0.001) # if change to async sleep, i don't see crash.
            yield AviaryModelResponse(generated_text="abcd")
        delta = time.time() - start
        print(f"**model throughput: {100 / delta}")


serve.run(Router.bind(SimpleGenerator.bind()))

time.sleep(3600)

```

locust.py

```
import requests
import time
import io


from locust import HttpUser, task, constant, events

class CodeGenClient(HttpUser):
    
    wait_time = constant(1)
    @task
    def send_serve_requests(self):
        request_meta = {
            "request_type": "InvokeEndpoint",
            "name": "Streamtest",
            "start_time": time.time(),
            "response_length": 0,
            "response": None,
            "context": {},
            "exception": None,
        }
        start_perf_counter = time.perf_counter()
        #r = self.client.get("/", stream=True)
        r = requests.get("http://localhost:8000", stream=True)
        if r.status_code != 200:
            print(r)
        else:
            for i, chunk in enumerate(r.iter_content(chunk_size=None, decode_unicode=True)):
                pass
            request_meta["response_time"] = (
                time.perf_counter() - start_perf_counter
            ) * 1000
            events.request.fire(**request_meta)
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Right now, future.result() returns and rasie an exception when future.cancel() is called although the coroutine itself is actually not cancelled. I verified this from a local example. 

This means the future.result can return and the task states are cleaned up (since task is cancelled) while we are running the actual coroutine & cpp code in a different thread. This causes a random sigabrt and sigsegv when there are lots of cancel & generator tasks are running.

For a generator, we are using a custom await between generator for the perf reason: https://github.com/ray-project/ray/blob/f11fa8422cec261d00d53bb6992e467ea1f45074/python/ray/_raylet.pyx#L1222. The cpp func running in a separte thread kept running while the task is cleaned up, which causes the segfault.

The same thing can technically happen for a regular async task (if it uses thread pool). If it doesn't use a thread pool, it doesn't happen because the future.cancel is not scheduled until the coroutine is awaited (whereas for a generator, it always have await between each run because of the threadpool above).

I fixed the issue by waiting until the coroutine is finished. Note that currently it is a blocking call, but it will mostly be a short time because when the cancellation error is raised the coroutine is already cancelled. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/39703

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by jjyao — APPROVED** [review]
```
Right now, future.result() returns and rasie an exception when future.cancel() is called although the coroutine itself is actually not cancelled. I verified this from a local example.
```

Is this behavior documented by Python?

**Comment by rkooo567** [comment]
It is not documented, but a relevant issue; https://github.com/python/cpython/issues/105836

**Comment by rkooo567** [comment]
I made some changes for more robust logic (the original logic has a blocking code, which is technically not triggered that often, but it is still bad).

1. Use asyncio.shield for `report_streaming_generator_output `. It ensures that when we run report_streaming_generator_output inside a thread pool, it is not cancelled.
2. Use `coroutine_complete_event ` to make sure we don't finish the task and clean up states before the coroutine `execute_streaming_generator_sync ` completes. 
3. We wait for `coroutine_complete_event` only when task is cancelled to minimize the perf impact. 
4. Instead of directly call future.cancel, call `eventloop.call_soon_threadsafe(future.cancel)`. I think this is probably the same behavior under the hood, but it is better being explicit.

File: locustfile.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import requests
```
Comment (edoakes) [inline locustfile.py:1]: remove?

File: python/ray/_raylet.pyx [hunk 2]
```diff
@@ -1196,36 +1196,43 @@ async def execute_streaming_generator_async(
 
     Args:
         context: The context to execute streaming generator.
+        coroutine_complete_event: The asyncio.Event to notify the
```
Comment (edoakes) [inline python/ray/_raylet.pyx:1199]: this and the type hint say asyncio.Event but a threading.Event is passed in

**Review by edoakes — CHANGES_REQUESTED** [review]
This looks extremely sketchy and blocking the loop makes me uncomfortable.

We are already using an event to signal the completion of the coroutine inside of `run_async_func_or_coro_in_event_loop`:
https://github.com/ray-project/ray/blob/9d4ae4eefba18e0abad1b6244403bb59d6226a86/python/ray/_raylet.pyx#L4298

I assume the current issue is that the callback added with `add_done_callback` is running prior to the coroutine actually completing. If that's the case, then how about we move `event.Notify()` into a `finally` block in the `async_func` just above?

I believe that'll accomplish the same thing without requiring the blocking `wait()` call.

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -2235,7 +2260,7 @@ cdef void cancel_async_task(
             function_descriptor, name_of_concurrency_group_to_execute)
         future = worker.core_worker.get_queued_future(task_id)
         if future is not None:
-            future.cancel()
+            eventloop.call_soon_threadsafe(future.cancel)
```
Comment (edoakes) [inline python/ray/_raylet.pyx:2263]: why's this change needed btw?

**Comment by edoakes** [comment]
@rkooo567 I just verified that my suggested change fixes the test case you've added. Without the following diff, I see the segfaults. With the following diff, the test passes consistently:
```
diff --git a/python/ray/_raylet.pyx b/python/ray/_raylet.pyx
index af1e777680..5449563478 100644
--- a/python/ray/_raylet.pyx
+++ b/python/ray/_raylet.pyx
@@ -4280,22 +4280,24 @@ cdef class CoreWorker:
             function_descriptor, specified_cgname)
 
         async def async_func():
-            if task_id:
-                async_task_id.set(task_id)
+            try:
+                if task_id:
+                    async_task_id.set(task_id)
 
-            if inspect.isawaitable(func_or_coro):
-                coroutine = func_or_coro
-            else:
-                coroutine = func_or_coro(*args, **kwargs)
+                if inspect.isawaitable(func_or_coro):
+                    coroutine = func_or_coro
+                else:
+                    coroutine = func_or_coro(*args, **kwargs)
 
-            return await coroutine
+                return await coroutine
+            finally:
+                event.Notify()
 
         future = asyncio.run_coroutine_threadsafe(async_func(), eventloop)
         if task_id:
             with self._task_id_to_future_lock:
                 self._task_id_to_future[task_id] = future
 
-        future.add_done_callback(lambda _: event.Notify())
         with nogil:
             (CCoreWorkerProcess.GetCoreWorker()
                 .YieldCurrentFiber(event))
```

**Comment by edoakes** [comment]
Made a PR to run CI against it: https://github.com/ray-project/ray/pull/40122

**Comment by rkooo567** [comment]
> I assume the current issue is that the callback added with add_done_callback is running prior to the coroutine actually completing. If that's the case, then how about we move event.Notify() into a finally block in the async_func just above?

@edoakes this fix makes sense

File: python/ray/_raylet.pyx (hunk 3 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx:2263]: This is to reduce the impact of the blocking call (so when things are cancelled, the coroutine is already context switched). I think with the new fix, it is not necessray

File: python/ray/_raylet.pyx (hunk 2 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx:1199]: ah I tried with non-blocking appraoch, and it didn't work, and I changed to blocking way. Let's just go with https://github.com/ray-project/ray/pull/40122 as it makes more sense to me..

**Comment by rkooo567** [comment]
there's a better/safer approach here; https://github.com/ray-project/ray/pull/40122

**Pull request closed without merging**